Blocks/MMQ/mmpq.h

```diff
@@ -70,83 +70,39 @@ class MMPQ {
     locks::SmartMutexLockGuard<MLS> lock(mutex_);
-    const auto timestamp = current::time::GetTimestampFromLockedSection(us);
-    if (!(timestamp > last_idx_ts_.us)) {
-      CURRENT_THROW(ss::InconsistentTimestampException(last_idx_ts_.us + std::chrono::microseconds(1), timestamp));
-    }
+    const auto us = current::time::GetTimestampFromLockedSection(std::forward<TIMESTAMP>(timestamp));
     ++last_idx_ts_.index;
-    last_idx_ts_.us = timestamp;
-    queue_.emplace(std::move(message), last_idx_ts_);
+    // Does not update `last_idx_ts_.us` at all. `UpdateHead()` must be called.
+    // This is to ensure the regular `Publish`, coming through the interface defined in `Blocks/SS/pubsub.h`,
+    // can publish into the future and utilize the full power of MMPQ.
+    queue_.emplace(std::forward<E>(entry), idxts_t(last_idx_ts_.index, us));
     condition_variable_.notify_all();
     return last_idx_ts_;
   }
 
-  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock, class T, typename US>
-  idxts_t DoPublishIntoTheFuture(T&& message, const US us) {
+  // NOTE: `UpdateHead` must be called for the entries to be processed, since otherwise, as entries can be published
+  // at non-increasing timestamps order, they must be held in the queue.
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock, typename TIMESTAMP>
+  void PublisherUpdateHeadImpl(TIMESTAMP&& timestamp) {
     locks::SmartMutexLockGuard<MLS> lock(mutex_);
-    const auto timestamp = current::time::GetTimestampFromLockedSection(us);
-    if (!(timestamp > last_idx_ts_.us)) {
-      CURRENT_THROW(ss::InconsistentTimestampException(last_idx_ts_.us + std::chrono::microseconds(1), timestamp));
+    const auto us = current::time::GetTimestampFromLockedSection(std::forward<TIMESTAMP>(timestamp));
+    if (!(us > last_idx_ts_.us)) {
+      CURRENT_THROW(ss::InconsistentTimestampException(last_idx_ts_.us + std::chrono::microseconds(1), us));
     }
-    ++last_idx_ts_.index;
-    // Don't update the timestamp.
-    queue_.emplace(std::move(message), idxts_t(last_idx_ts_.index, timestamp));
+    last_idx_ts_.us = us;
     condition_variable_.notify_all();
-    return last_idx_ts_;
   }
 
-  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock, typename US>
-  void DoUpdateHead(const US us) {
-    locks::SmartMutexLockGuard<MLS> lock(mutex_);
-    const auto timestamp = current::time::GetTimestampFromLockedSection(us);
-    if (!(timestamp > last_idx_ts_.us)) {
-      CURRENT_THROW(ss::InconsistentTimestampException(last_idx_ts_.us + std::chrono::microseconds(1), timestamp));
-    }
-    last_idx_ts_.us = timestamp;
-    condition_variable_.notify_all();
-  }
+ private:
+  MMPQImpl(const MMPQImpl&) = delete;
+  MMPQImpl(MMPQImpl&&) = delete;
+  void operator=(const MMPQImpl&) = delete;
+  void operator=(MMPQImpl&&) = delete;
```

Comment:
Yes. My line of thought was that it'd eventually be `Owned<>`, which is both more universal usage-wise and less code maintenance-wise. What do you think?